OpenRA.Game/WorldUtils.cs

```diff
@@ -41,6 +41,116 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+			var overselect = dir * (1 + ((int)rayWidth.Length / 1024));
+			var finalTarget = target + overselect;
+			var finalSource = source - overselect;
+
+			var actorsInSquare = world.ActorMap.ActorsInBox(finalTarget, finalSource);
+			var intersectedActors = new List<Actor>();
+
+			foreach (var currActor in actorsInSquare)
+			{
+				long actorWidth;
+				var healthInfo = currActor.Info.TraitInfoOrDefault<HealthInfo>();
+				if (healthInfo != null)
+				{
+					actorWidth = healthInfo.Radius.Length;
+				}
+				else
+					actorWidth = 0;
+
+				var distance = MinimumDistancePointAndVector(target, source, currActor.CenterPosition);
+				var maxReach = (long)(actorWidth + rayWidth.Length);
+				maxReach *= maxReach;
+				if (distance <= maxReach)
+				{
+					intersectedActors.Add(currActor);
+				}
+			}
+
+			return intersectedActors;
+		}
+
+		/// <summary>
+		/// Returns the (squared) minimum distance between a line and a point.
+		/// From http://stackoverflow.com/questions/849211/shortest-distance-between-a-point-and-a-line-segment
+		/// To avoid floats, it returns the square of the distance.
+		/// </summary>
+		/// <param name="target">The target point (head) of the line</param>
+		/// <param name="source">The source point (tail) of the line</param>
+		/// <param name="point">The point that the minimum distance should be found to</param>
+		/// <returns>The square of the minimum distance between the line and point.</returns>
+		private static long MinimumDistancePointAndVector(WPos target, WPos source, WPos point)
+		{
+			// Return minimum distance between line segment vw and point p
+			var squaredLength = SquaredDistance(target, source);
+
+			// target = source case, so just find distance from the point to the target.
+			if (squaredLength == 0.0) return SquaredDistance(target, point);
+
+			// Consider the line extending the segment, parameterized as target + t (source - target).
+			// We find projection of point onto the line.
+			// It falls where t = [(point - target) . (source - target)] / |source - target|^2
+			double t = DotProduct(target, source, point);
+
+			// Beyond the 'target' end of the segment
+			if (t < 0.0)
+				return SquaredDistance(point, target);
+
+			// Beyond the 'source end of the segment
+			if (t > 1.0)
+				return SquaredDistance(point, source);
+
+			// Projection falls on the segment
+			var newX = (int)((double)target.X + t * (double)(source.X - target.X));
+			var newY = (int)((double)target.Y + t * (double)(source.Y - target.Y));
+			var projection = new WPos(newX, newY, 0);
+			return SquaredDistance(point, projection);
+		}
+
+		private static long SquaredDistance(WPos target, WPos source)
+		{
+			return ((target.X - source.X) * (target.X - source.X)) + ((target.Y - source.Y) * (target.Y - source.Y));
+		}
+
+		private static double DotProduct(WPos target, WPos source, WPos point)
+		{
+			double xDiff = (double)((point.X - target.X) * (source.X - target.X));
```

Comment:
As suggested above, you can rewrite this by dropping the double cast and multiplying these components by 1024 and returning an int.